File: 树/103.二叉树的锯齿型层序遍历.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        queue = deque([root])
        flag = 1     # 1表示从左往右 -1表示从右往左
        res = []
        while queue:
            layerLen = len(queue)
            layer = []
            for _ in range(layerLen):
                if flag == 1: # 从左往右入队
                    cur = queue.popleft()
                    layer.append(cur.val)
                    if cur.left:
                        queue.append(cur.left)
                    if cur.right:
                        queue.append(cur.right)
                else: # 从右往左入队
                    cur = queue.pop()
                    layer.append(cur.val)
                    if cur.right:
                        queue.appendleft(cur.right)
                    if cur.left:
                        queue.appendleft(cur.left)
            res.append(layer)
            flag *= -1
        return res

File: 树/103.二叉树的锯齿型层序遍历.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}  // NOTE: 数字
 */
var zigzagLevelOrder = function (root) {
  if (!root) return []; //! 记得处理特殊情况
  const queue = [];
  const res = [];
  queue.push(root);
  let isLeftToRight = true; //! flag声明在遍历外层
  while (queue.length) {
    const layerLen = queue.length;
    const layer = [];
    for (let i = 0; i < layerLen; i++) {
      const node = queue.shift(); //! 这里是先进先出
      // 添加当前节点子节点
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
      // 通过遍历方向确定是前加还是后加
      if (isLeftToRight) {
        layer.push(node.val); //NOTE: 注意返回值要求
      } else {
        layer.unshift(node.val);
      }
    }
    isLeftToRight = !isLeftToRight; //! 每一层结束之后改变遍历方向
    res.push(layer);
  }
  return res;
};

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function (root) {
  // 处理根节点为空的情况
  if (!root) return [];

  const queue = [];
  const res = [];
  queue.push(root);
  let isLeftToRight = true;

  while (queue.length) {
    const layerLen = queue.length;
    const layer = [];

    for (let i = 0; i < layerLen; i++) {
      // 从队列头部取出元素
      const node = queue.shift();

      // 添加当前节点子节点
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);

      // 根据遍历方向记录节点值
      if (isLeftToRight) {
        layer.push(node.val);
      } else {
        layer.unshift(node.val);
      }
    }

    // 每一层结束后改变遍历方向
    isLeftToRight = !isLeftToRight;
    res.push(layer);
  }

  return res;
};
